### PR TITLE
update swagger on build

### PR DESCRIPTION
**Pull request by esmadau**
## Description
- Update swagger on building sln
- Added pipeline step to check for latest swagger changes and, if any detected, fail pipeline

[Example of failure](https://microsofthealthoss.visualstudio.com/DicomServer/_build/results?buildId=22296&view=logs&j=1b246169-45e3-528c-e738-c7727e58b894&t=ad1c6c01-420c-5688-4964-9ad6b003e95b&l=62)
[Example of success](https://microsofthealthoss.visualstudio.com/DicomServer/_build/results?buildId=22303&view=logs&j=1b246169-45e3-528c-e738-c7727e58b894&t=ad1c6c01-420c-5688-4964-9ad6b003e95b&l=33)

## Related issues
Addresses issue [[AB#93981](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/93981)], which is a step towards
- having up to date swagger files
- failing ADO pipeline when detecting breaking API changes in swagger

## Testing
Tested locally using the script committed here as well as using pipeline.

### CODE REVIEW

File: build/common/scripts/CheckForSwaggerChanges.ps1 [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    [string]$SwashbuckleCLIVersion = '6.4.0'
+)
+$ErrorActionPreference = 'Stop'
+dotnet new tool-manifest --force
+dotnet tool install --version $SwashbuckleCLIVersion Swashbuckle.AspNetCore.Cli
+
+Write-Host "Using swagger version ..."
+dotnet tool list | Select-String "swashbuckle"
+
```
Comment (wsugarman) [inline build/common/scripts/CheckForSwaggerChanges.ps1:29]: nit: extra line

File: build/common/scripts/CheckForSwaggerChanges.ps1 [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    Write-Host "Comparing generated swagger with what was checked in ..."
+    $HasDifferences = (Compare-Object -ReferenceObject (Get-Content -Path $WritePath) -DifferenceObject (Get-Content -Path ".\swagger\$Version\swagger.yaml"))
+    if ($HasDifferences){
+        Write-Host $HasDifferences
+        throw "The swagger yaml checked in with this PR is not up to date with code. Please build the sln, which will trigger a hook to autogenerate these files on your behalf. Differences shown above."
+    }else{
+        Write-Host "Swagger checked in with this PR is up to date with code."
+    }
+
```
Comment (wsugarman) [inline build/common/scripts/CheckForSwaggerChanges.ps1:49]: nit: extra new lines

File: build/common/scripts/CheckForSwaggerChanges.ps1 [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    $WritePath=(Join-Path -Path "$SwaggerDir" -ChildPath "$Version.yaml")
+
+    dotnet swagger tofile --yaml --output $WritePath "$AssemblyDir" $Version
+
+    Write-Host "Comparing generated swagger with what was checked in ..."
+    $HasDifferences = (Compare-Object -ReferenceObject (Get-Content -Path $WritePath) -DifferenceObject (Get-Content -Path ".\swagger\$Version\swagger.yaml"))
+    if ($HasDifferences){
+        Write-Host $HasDifferences
+        throw "The swagger yaml checked in with this PR is not up to date with code. Please build the sln, which will trigger a hook to autogenerate these files on your behalf. Differences shown above."
+    }else{
```
Comment (wsugarman) [inline build/common/scripts/CheckForSwaggerChanges.ps1:46]: nit: space before `else`

File: build/common/scripts/CheckForSwaggerChanges.ps1 [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+foreach ($Version in $Versions)
+{
+    Write-Host "Generating swagger yaml file for $Version"
+    $WritePath=(Join-Path -Path "$SwaggerDir" -ChildPath "$Version.yaml")
+
+    dotnet swagger tofile --yaml --output $WritePath "$AssemblyDir" $Version
+
+    Write-Host "Comparing generated swagger with what was checked in ..."
+    $HasDifferences = (Compare-Object -ReferenceObject (Get-Content -Path $WritePath) -DifferenceObject (Get-Content -Path ".\swagger\$Version\swagger.yaml"))
```
Comment (wsugarman) [inline build/common/scripts/CheckForSwaggerChanges.ps1:42]: Yeah, I think this makes sense. We always want the checked-in file to match the latest spec

File: src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj [hunk 5]
```diff
@@ -32,4 +32,10 @@
     <Folder Include="wwwroot\" />
   </ItemGroup>
 
+  <Target Name="SwaggerPostBuildTarget" AfterTargets="Build" Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:35]: You can use the condition `'$(ContinuousIntegrationBuild)' != 'true'`. We only pass `ContinuousIntegrationBuild` in our CI pipelines as it controls a couple settings

File: src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj [hunk 6]
```diff
@@ -32,4 +32,10 @@
...
+    <Exec Command="dotnet tool restore"></Exec>
+    <Exec Command="dotnet swagger tofile  --yaml --output ..\..\swagger\v1\swagger.yaml bin\x64\Debug\net6.0\Microsoft.Health.Dicom.Web.dll v1"></Exec>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:37]: There should be a property like `$(OutputPath)` for `bin\x64\Debug\net6.0\`

Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:37]: How are we ensuring users actually have this tool locally?

Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:37]: Is there a way to unify this and the CI script? Could we instead always generate the files?

File: src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj [hunk 7]
```diff
@@ -32,4 +32,10 @@
     <Folder Include="wwwroot\" />
   </ItemGroup>
 
+  <Target Name="SwaggerPostBuildTarget" AfterTargets="Build" Condition="'$(ContinuousIntegrationBuild)' != 'true'">
+    <Exec Command="dotnet tool restore"></Exec>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:36]: If we're going to use restore, then we can skip any description in a README. Let's instead commit a tool "manifest" file and specify it here in the `dotnet tool restore` command.

**Review by jovinson-ms — DISMISSED** [review]

File: src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj (hunk 7 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:36]: This way we can standardize our tooling across msbuild, docker, and the CI script

File: .config/dotnet-tools.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  "isRoot": true,
+  "tools": {
+    "swashbuckle.aspnetcore.cli": {
+      "version": "6.4.0",
+      "commands": [
+        "swagger"
+      ]
+    }
+  }
+}
```
Comment (wsugarman) [inline .config/dotnet-tools.json:12]: nit: can you add a new line at the end?

File: build/common/scripts/CheckForSwaggerChanges.ps1 [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    [string]$SwaggerDir,
+
+    [string]$AssemblyDir,
+
+    [String[]]$Versions,
+
+    [string]$SwashbuckleCLIVersion = '6.4.0'
+)
+$ErrorActionPreference = 'Stop'
+dotnet new tool-manifest --force
```
Comment (wsugarman) [inline build/common/scripts/CheckForSwaggerChanges.ps1:24]: You can now restore the existing manifest instead

File: src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj (hunk 7 above)
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Web/Microsoft.Health.Dicom.Web.csproj:36]: Do we need to manually specify the manifest or will it find it automatically based on convention?

**Review by wsugarman — APPROVED** [review]

**Pull request merged**